### PR TITLE
Plugins: bundle plugins

### PR DESCRIPTION
**Pull request by ryantxu**
This PR is an attempt to have grafana core support bundling internal and external plugins.  It will let us use the standard plugin infrastructure (SystemJS) for some core plugins and allow a more composable final package.

This PR just builds the plugins and and adds them to the local scan path -- a future task should clean up the distribution before packaging them.

### CODE REVIEW

**Comment by dprokop** [comment]
Looking for some more details about the idea behind this work @ryantxu :)

**Comment by torkelo** [comment]
The main idea is to have a manifest file, https://github.com/grafana/grafana/pull/20850/files#diff-6d9eccf8db3e07f6315895fc7106a5a1R3 

That lists both internal & external plugins, then have a script that can handle fetching and building those (and dev watching for internal). The only difference between internal vs external in this context is internal in this main repo external is plugin in outside repo. 

The point of this is then to make this fetch & build step part of the package build scripts so we can bundle all these plugins in package. We also need grafana to scan plugins in this bundled plugins folder. And some logic that checks if a newer version of a bundled plugin is installed in an external plugin folder then use that.

File: pkg/plugins/plugins.go [hunk 1]
```diff
@@ -72,6 +72,14 @@ func (pm *PluginManager) Init() error {
 		return errutil.Wrapf(err, "Failed to scan main plugin directory '%s'", plugDir)
 	}
 
+	pm.log.Info("Checking managed plugins")
+	plugDir = path.Join(setting.HomePath, "plugins-managed")
```
Comment (ryantxu) [inline pkg/plugins/plugins.go]: @torkelo -- thoughts on the actual path name?  maybe `managed-plugins` (I did it as plugins-managed so it shows up next to the plugins source tree)

File: plugins/app/example-app/src/legacy/angular_example_page.ts [hunk 2]
```diff
@@ -1,10 +1,8 @@
-import { GrafanaRootScope } from 'app/routes/GrafanaCtrl';
-
 export class AngularExamplePageCtrl {
   static templateUrl = 'legacy/angular_example_page.html';
 
   /** @ngInject */
-  constructor($scope: any, $rootScope: GrafanaRootScope) {
+  constructor($scope: any, $rootScope: any) {
```
Comment (ryantxu) [inline plugins/app/example-app/src/legacy/angular_example_page.ts]: GrafanaRootScope is not exposed to plugins, so just using any

File: pkg/plugins/plugins.go (hunk 1 above)
Comment (torkelo) [reply pkg/plugins/plugins.go]: Think I like plugins-bundled or plugins-bundle  more

File: package.json [hunk 3]
```diff
@@ -270,7 +270,8 @@
   },
   "workspaces": {
     "packages": [
-      "packages/*"
+      "packages/*",
+      "plugins-bundled/internal/*"
```
Comment (ryantxu) [inline package.json]: just making stuff up!  but this must do something!

File: plugins-bundled/internal/input-datasource/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "extends": "../../../packages/grafana-toolkit/src/config/tsconfig.plugin.json",
```
Comment (ryantxu) [inline plugins-bundled/internal/input-datasource/tsconfig.json]: I tested locally with the only difference that this path is:
```
"extends": "./node_modules/@grafana/toolkit/src/config/tsconfig.plugin.json",
```
and everything works

File: public/app/features/plugins/built_in_plugins.ts [hunk 5]
```diff
@@ -54,8 +52,6 @@ import * as barGaugePanel from 'app/plugins/panel/bargauge/module';
 import * as logsPanel from 'app/plugins/panel/logs/module';
 import * as newsPanel from 'app/plugins/panel/news/module';
 
-const exampleApp = async () => await import(/* webpackChunkName: "exampleApp" */ 'app/plugins/app/example-app/module');
```
Comment (ryantxu) [inline public/app/features/plugins/built_in_plugins.ts]: Deleting the app and we can maintain it from:
https://github.com/grafana/simple-app-plugin

**Comment by ryantxu** [comment]
@dprokop -- can I get some help/guidance on this?  I am not sure how to make the packages within `"plugins-bundled/internal/*"` build

**Comment by dprokop** [comment]
@ryantxu yeah, I'm struggling with this as well. Basically running toolkit task from within this directory results in typecheck errors agains grafana-ui (slate react typings + no implicit any on files related to those). Spent some time on this already but haven't found a solution yet. Would love to ignore @grafana/* packages typecheck, but couldn't make it work.

A quick "workaround" would be NOT to treat those bundled plugins as workspaces (that's something we don't want to do anyways I think). Then we would have to install "managed" plugin's dependencies and rely on installed grafana-toolkit instead of the one from workspaces. I assume we gonna guarantee somehow that the fetched plugin is using the packages versions that are in line with the target Grafana version?

File: package.json (hunk 3 above)
Comment (torkelo) [reply package.json]: why do we need this? bundled plugins should not contain npm packages that we import from other packages / core

**Comment by dprokop** [comment]
@ryantxu I think you should be able now to bundle the plugins within `plugins-bundled` directory (90904ccb011cb77a29a8f94ff8629baae156f1e1 takes care of filtering the noise not related to the plugin). 

I haven't figured out yet why typecheck on such packages fails for grafana-ui dependency. There is a similar issue described here https://stackoverflow.com/questions/54288910/ignore-ts-errors-in-project-with-yarn-workspaces but haven't found a solution for it now.

Running tyopecheck with fork-ts-checker-webpack-plugin allows scoping the subject of typecheck results, so we get can get only the results that are important from plugin's POV.

In 0b42b2eac889e2c0a16d15dbeb72408658a0bc8c I've added a draft for the toolkit task for bundling the plugins. For now it performs only yarn dev on plugins in `plugins-bundled/[interal|external]`

**Comment by ryantxu** [comment]
This is looking good -- thank you @dprokop 

It looks like the test are struggling though when I run `yarn dev` in a plugin I get:
image

**Comment by torkelo** [comment]
regarding typescript or other issues, since these will be sub "npm / node" projects (ie in subfolder under the main one), could it be that we will run into problems with node_modules hoisting, ie that node will take dependencies from the root grafana project instead? try removing top repo root node_modules folder see if that helps. 

I could not find away to disable hoisting, so if could be that we will have issues with that and that we will need to move the root package.json / yarn files to the public dir to solve it.

**Comment by dprokop** [comment]
> regarding typescript or other issues, since these will be sub "npm / node" projects (ie in subfolder under the main one), could it be that we will run into problems with node_modules hoisting, ie that node will take dependencies from the root grafana project instead? try removing top repo root node_modules folder see if that helps.

That's true. In proposed implementation plugins are treated as workspaces meaning they are subject of hoisting. At the same time, dev wise, we will have to run `npm install` per each plugin so that each plugin have it's own set of dependencies. And this leads to `@grafana/*` packages being installed from npm as well. Not sure what to do with this. Maybe we could link local dependencies, but then everyone would have to have those linked locally which is sub-optimal

**Comment by dprokop** [comment]
quiet, stale!

**Comment by dprokop** [comment]
not stale

**Comment by torkelo** [comment]
Is there a lot of remaining work to make this mergable?  And continue work in follow-up PRs

**Comment by torkelo** [comment]
Maybe create an epic & todo if there is a lot of small remaining tasks

File: package.json (hunk 3 above)
Comment (ryantxu) [reply package.json:286]: @dprokop ?  is this necessary?  would not be used for npm/webpack, only vscode

**Comment by ryantxu** [comment]
@aknuds1 -- this is the first step in shipping grafana with a set of managed plugins.  Essentially it is an extra folder that gets scanned.  After this PR, we can look at adding other plugins to the "standard" distribution

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Just a teeny tiny suggestion. Otherwise looks good to me.

**Review by aknuds1 — APPROVED** [review]
LGTM

**Pull request merged**